aggsender/flows/flow_base.go

```diff
@@ -573,20 +614,140 @@ func (f *baseFlow) getLastSentBlockAndRetryCount(lastSentCertificateInfo *types.
+	//    unclaims before the current block
+	cutBlock, err := f.validateUnclaimsForUnfinalizedGERs(certParams, cache)
+	if err != nil {
+		return nil, fmt.Errorf("error validating unclaims for unfinalized GERs: %w", err)
+	}
+	if cutBlock != 0 {
+		if cutBlock-1 < certParams.FromBlock {
+			return nil, fmt.Errorf(
+				"cannot create certificate: claim at block %d (start block %d) cannot be included and no valid blocks before it",
+				cutBlock, certParams.FromBlock)
 		}
+		f.log.Warnf("found claim with unclaim after later unfinalized claim at block %d, cutting certificate at block %d",
+			cutBlock, cutBlock-1)
+		return certParams.AdjustToBlock(cutBlock - 1)
 	}
 
 	return certParams, nil
 }
+
+// validateUnclaimsForUnfinalizedGERs validates unclaims for unfinalized GERs that don't exist on L1
+// in a single pass. This function combines two checks:
+//  1. Checks if any claim with unfinalized GER that doesn't exist on L1 has an unclaim that appears
+//     after a later unfinalized claim. If so, returns the block number to cut at.
+//  2. Validates that any previous claims with unfinalized GERs that don't exist on L1 have their
+//     unclaims before the current claim's block. If a claim without an unclaim is found, returns
+//     the block number to cut at.
+//
+// Returns the earliest cut block found (or 0 if no cut is needed) and an error if validation fails.
+func (f *baseFlow) validateUnclaimsForUnfinalizedGERs(
+	certParams *types.CertificateBuildParams,
+	cache *gerStatusCache) (uint64, error) {
+	// Build a map of unclaims by GlobalIndex for quick lookup
+	unclaimMap := make(map[string]uint64)
+	for _, unclaim := range certParams.Unclaims {
+		key := unclaim.GlobalIndex.String()
+		// Keep the earliest unclaim if there are multiple
+		if existing, ok := unclaimMap[key]; !ok || unclaim.BlockNumber < existing {
+			unclaimMap[key] = unclaim.BlockNumber
+		}
+	}
+
+	var earliestCutBlock uint64
+
+	// Single pass through all claims to perform both checks
+	for i, claim := range certParams.Claims {
+		// Check if this claim's GER is finalized
+		isGERFinalized, err := f.getGERFinalizedStatus(cache, claim.GlobalExitRoot, certParams.L1InfoTreeLeafCount)
+		if err != nil {
+			return 0, fmt.Errorf("error checking if claim's GER %s is finalized: %w",
+				claim.GlobalExitRoot.String(), err)
+		}
+
+		// Skip if GER is finalized
+		if isGERFinalized {
+			continue
+		}
+
+		// Check 1: Validate that claims before currentBlockNum with unfinalized GERs that don't exist on L1
+		// have their unclaims before currentBlockNum
+		unclaimBlock, hasUnclaim := unclaimMap[claim.GlobalIndex.String()]
+		if !hasUnclaim {
+			// Found a claim without an unclaim - this is a cut point
+			if earliestCutBlock == 0 || claim.BlockNum < earliestCutBlock {
+				earliestCutBlock = claim.BlockNum
+			}
+			return earliestCutBlock, nil
+		}
+
+		// Check 2: Check if this claim's unclaim appears after any later unfinalized claim
```

Comment:
The logic looks correct, but the comment could be a more descriptive:

```suggestion
// Check 2: Ensure we can include this claim's unclaim without being forced to include
// a later unfinalized claim that doesn't exist on L1.
// We need to cut certificate if:
// - a later unfinalized claim (that doesn't exist on L1) appears at or before our unclaim block
// - and that later claim either has no unclaim OR has an unclaim after our unclaim
```